migen/test/test_platform.py

```diff
@@ -33,14 +38,53 @@ def __init__(self, plat):
 
 class TestExamplesPlatform(unittest.TestCase):
     def test_platforms(self):
+        def mkdir_and_build_source(plat, mod, name):
+            # Test should not fail if toolchain doesn't exist,
+            # which will happen if source=True (the default).
+            curr_exc = None
+            for source in (True, False):
+                cwd = os.getcwd()
+                try:
+                    mkdir_and_build(plat, name, source=source)
+                    break
+                except FileNotFoundError as e:
+                    # Platform will be finalized during first run,
+                    # so get a fresh instance for both iterations.
+                    plat = importlib.import_module(mod).Platform()
+                    curr_exc = e
+                    continue
+                finally:
+                    # If build fails, make sure we return to our
+                    # original directory; mkdir_and_build() will
+                    # purge the build_dir as part of cleanup.
+                    os.chdir(cwd)
```

Comment:
Really needed?

Reply:
Toolchain classes don't restore the previous working directory if they throw an exception (nor should they really IMO; in cases besides tests, throwing an exception in build should be fatal). The `tempfile.TemporaryDirectory` context will delete the build directory as part of `__exit__`. 

If a platform fails to build when `source=True`, there will be an exception thrown from within `plat.build`, and then a second attempt to call `plat.build` with `source=False`. Without this line, the next call to build will try run `os.getcwd()` in a nonexistent directory, which throws an exception.